File: android/src/main/kotlin/com/cloudwebrtc/webrtc/controller/PeerConnectionController.kt
package com.cloudwebrtc.webrtc.controller

import com.cloudwebrtc.webrtc.model.*
import com.cloudwebrtc.webrtc.proxy.MediaStreamTrackProxy
import com.cloudwebrtc.webrtc.proxy.PeerConnectionProxy
import com.cloudwebrtc.webrtc.proxy.RtpTransceiverProxy
import com.cloudwebrtc.webrtc.utils.AnyThreadSink
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.runBlocking

/**
 * Controller for the [PeerConnectionProxy] functional.
 *
 * @property messenger messenger used for creating new [MethodChannel]s.
 * @property peer underlying [MediaStreamTrackProxy] on which method calls will be performed.
 */
class PeerConnectionController(
    private val messenger: BinaryMessenger,
    private val peer: PeerConnectionProxy
) :
    MethodChannel.MethodCallHandler, EventChannel.StreamHandler, IdentifiableController {
    /**
     * Unique ID of the [MethodChannel] of this controller.
     */
    private val channelId = nextChannelId()

    /**
     * Channel which will be listened for the [MethodCall]s.
     */
    private val methodChannel: MethodChannel =
        MethodChannel(messenger, ChannelNameGenerator.withId("PeerConnection", channelId))

    /**
     * Event channel into which all [PeerConnectionProxy] events will be sent.
     */
    private val eventChannel: EventChannel =
        EventChannel(messenger, ChannelNameGenerator.withId("PeerConnectionEvent", channelId))

    /**
     * Event sink into which all [PeerConnectionProxy] events will be sent.
     */
    private var eventSink: AnyThreadSink? = null

    /**
     * [PeerConnectionProxy] events observer which will send all events to the [eventSink].
     */
    private val eventObserver = object : PeerConnectionProxy.Companion.EventObserver {
        override fun onTrack(track: MediaStreamTrackProxy, transceiver: RtpTransceiverProxy) {
            eventSink?.success(
                mapOf(
                    "event" to "onTrack",
                    "track" to MediaStreamTrackController(messenger, track).asFlutterResult(),
                    "transceiver" to RtpTransceiverController(
                        messenger,
                        transceiver
                    ).asFlutterResult()
                )
            )
        }

        override fun onIceConnectionStateChange(iceConnectionState: IceConnectionState) {
            eventSink?.success(
                mapOf(
                    "event" to "onIceConnectionStateChange",
                    "state" to iceConnectionState.value
                )
            )
        }

        override fun onSignalingStateChange(signalingState: SignalingState) {
            eventSink?.success(
                mapOf(
                    "event" to "onSignalingStateChange",
                    "state" to signalingState.value
                )
            )
        }

        override fun onConnectionStateChange(peerConnectionState: PeerConnectionState) {
            eventSink?.success(
                mapOf(
                    "event" to "onConnectionStateChange",
                    "state" to peerConnectionState.value
                )
            )
        }

        override fun onIceGatheringStateChange(iceGatheringState: IceGatheringState) {
            eventSink?.success(
                mapOf(
                    "event" to "onIceGatheringStateChange",
                    "state" to iceGatheringState.value
                )
            )
        }

        override fun onIceCandidate(candidate: IceCandidate) {
            eventSink?.success(
                mapOf(
                    "event" to "onIceCandidate",
                    "candidate" to candidate.intoMap()
                )
            )
        }

        override fun onNegotiationNeeded() {
            eventSink?.success(
                mapOf(
                    "event" to "onNegotiationNeeded"
                )
            )
        }
    }

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
        peer.addEventObserver(eventObserver)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "createOffer" -> {
                runBlocking {
                    result.success(peer.createOffer().intoMap())
                }
            }
            "createAnswer" -> {
                runBlocking {
                    result.success(peer.createAnswer().intoMap())
                }
            }
            "setLocalDescription" -> {
                val descriptionArg: Map<String, Any>? = call.argument("description")
                val description = if (descriptionArg == null) {
                    null
                } else {
                    SessionDescription.fromMap(descriptionArg)
                }
                runBlocking {
                    peer.setLocalDescription(description)
                }
                result.success(null)
            }
            "setRemoteDescription" -> {
                val descriptionArg: Map<String, Any> = call.argument("description")!!
                runBlocking {
                    peer.setRemoteDescription(SessionDescription.fromMap(descriptionArg))
                }
                result.success(null)
            }
            "addIceCandidate" -> {
                val candidate: Map<String, Any> = call.argument("candidate")!!
                runBlocking {
                    peer.addIceCandidate(IceCandidate.fromMap(candidate))
                }
                result.success(null)
            }
            "addTransceiver" -> {
                val mediaType = MediaType.fromInt(call.argument("mediaType")!!)
                val transceiverInitArg: Map<String, Any>? = call.argument("init")
                val transceiver = if (transceiverInitArg == null) {
                    peer.addTransceiver(mediaType, null)
                } else {
                    peer.addTransceiver(mediaType, RtpTransceiverInit.fromMap(transceiverInitArg))
                }
                val transceiverController = RtpTransceiverController(messenger, transceiver)
                result.success(transceiverController.asFlutterResult())
            }
            "getTransceivers" -> {
                result.success(
                    peer.getTransceivers()
                        .map { RtpTransceiverController(messenger, it).asFlutterResult() })
            }
            "restartIce" -> {
                peer.restartIce()
                result.success(null)
            }
            "dispose" -> {
                dispose()
                result.success(null)
            }
        }
    }

    override fun onListen(obj: Any?, sink: EventChannel.EventSink?) {
        if (sink != null) {
            eventSink = AnyThreadSink(sink)
        }
    }

    override fun onCancel(obj: Any?) {
        eventSink = null
    }

    /**
     * Converts this [PeerConnectionController] to the Flutter's method call result.
     *
     * @return [Map] generated from this controller which can be returned to the Flutter side.
     */
    fun asFlutterResult(): Map<String, Any> = mapOf<String, Any>(
        "channelId" to channelId,
        "id" to peer.id
    )

    /**
     * Closes method and event channels of this [PeerConnectionController].
     *
     * Disposes underlying [PeerConnectionProxy].
     */
    private fun dispose() {
        methodChannel.setMethodCallHandler(null)
        peer.removeEventObserver(eventObserver)
        peer.dispose()
        eventChannel.setStreamHandler(null)
        eventSink?.endOfStream()
    }
}

File: ios/Classes/controller/PeerConnectionController.swift
import Flutter

/// Controller for the `PeerConnection`.
class PeerConnectionController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of the proxy of PeerConnection.
  private var peer: PeerConnectionProxy

  /// ID of channels created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Controller for the `eventChannel` management.
  private var eventController: EventController

  /// Event channel for communicating with Flutter side.
  private var eventChannel: FlutterEventChannel

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Creates new `PeerConnectionController` for the provided `PeerConnectionProxy`.
  init(messenger: FlutterBinaryMessenger, peer: PeerConnectionProxy) {
    let channelName = ChannelNameGenerator.name(name: "PeerConnection", id: self.channelId)
    self.eventController = EventController()
    self.messenger = messenger
    self.peer = peer
    self.peer.addEventObserver(
      eventObserver: PeerEventController(
        messenger: self.messenger, eventController: self.eventController))
    self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    self.eventChannel = FlutterEventChannel(
      name: ChannelNameGenerator.name(name: "PeerConnectionEvent", id: self.channelId),
      binaryMessenger: messenger)
    self.channel.setMethodCallHandler({ (call, result) in
      try! self.onMethodCall(call: call, result: result)
    })
    self.eventChannel.setStreamHandler(self.eventController)
  }

  /// Handles all supported Flutter method calls for the `PeerConnectionProxy`.
  func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "createOffer":
      Task {
        let sdp = try! await self.peer.createOffer()
        result(sdp.asFlutterResult())
      }
    case "createAnswer":
      Task {
        let sdp = try! await self.peer.createAnswer()
        result(sdp.asFlutterResult())
      }
    case "setLocalDescription":
      let description = argsMap!["description"] as? [String: Any]
      let type = description!["type"] as? Int
      let sdp = description!["description"] as? String
      Task {
        var desc: SessionDescription?
        if sdp == nil {
          desc = nil
        } else {
          desc = SessionDescription(
            type: SessionDescriptionType(rawValue: type!)!, description: sdp!)
        }
        try! await self.peer.setLocalDescription(description: desc)
        result(nil)
      }
    case "setRemoteDescription":
      let descriptionMap = argsMap!["description"] as? [String: Any]
      let type = descriptionMap!["type"] as? Int
      let sdp = descriptionMap!["description"] as? String
      Task {
        try! await self.peer.setRemoteDescription(
          description: SessionDescription(
            type: SessionDescriptionType(rawValue: type!)!, description: sdp!))
        result(nil)
      }
    case "addIceCandidate":
      let candidateMap = argsMap!["candidate"] as? [String: Any]
      let sdpMid = candidateMap!["sdpMid"] as? String
      let sdpMLineIndex = candidateMap!["sdpMLineIndex"] as? Int
      let candidate = candidateMap!["candidate"] as? String
      Task {
        try! await self.peer.addIceCandidate(
          candidate: IceCandidate(
            sdpMid: sdpMid!, sdpMLineIndex: sdpMLineIndex!, candidate: candidate!))
        result(nil)
      }
    case "addTransceiver":
      let mediaType = argsMap!["mediaType"] as? Int
      let initArgs = argsMap!["init"] as? [String: Any]
      let direction = initArgs!["direction"] as? Int
      let transceiverInit = TransceiverInit(direction: TransceiverDirection(rawValue: direction!)!)
      let transceiver = RtpTransceiverController(
        messenger: self.messenger,
        transceiver: self.peer.addTransceiver(
          mediaType: MediaType(rawValue: mediaType!)!, transceiverInit: transceiverInit))
      result(transceiver.asFlutterResult())
    case "getTransceivers":
      result(
        self.peer.getTransceivers().map {
          RtpTransceiverController(messenger: self.messenger, transceiver: $0).asFlutterResult()
        })
    case "restartIce":
      self.peer.restartIce()
      result(nil)
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      self.eventChannel.setStreamHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller to the Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    return [
      "channelId": self.channelId,
      "id": self.peer.getId(),
    ]
  }
}
